qiskit/aqua/operators/primitive_ops/pauli_sum_op.py

```diff
@@ -0,0 +1,405 @@
+
+
+class PauliSumOp(PrimitiveOp):
+    """Class for Operators backend by Terra's ``SparsePauliOp`` class."""
+
+    def __init__(
+            self,
+            primitive: SparsePauliOp,
+            coeff: Union[int, float, complex, ParameterExpression] = 1.0,
+    ) -> None:
+        """
+        Args:
+            primitive: The SparsePauliOp which defines the behavior of the underlying function.
+            coeff: A coefficient multiplying the primitive.
+
+        Raises:
+            TypeError: invalid parameters.
+        """
+        if not isinstance(primitive, SparsePauliOp):
+            raise TypeError(
+                "PauliSumOp can only be instantiated with SparsePauliOp, "
+                f"not {type(primitive)}"
+            )
+
+        super().__init__(primitive, coeff=coeff)
+
+    def primitive_strings(self) -> Set[str]:
+        return {"SparsePauliOp"}
+
+    @property
+    def num_qubits(self) -> int:
+        return self.primitive.num_qubits  # type: ignore
+
+    def add(self, other: OperatorBase) -> OperatorBase:
+        if not self.num_qubits == other.num_qubits:
+            raise ValueError(
+                "Sum over operators with different numbers of qubits, {} and {}, is not well "
+                "defined".format(self.num_qubits, other.num_qubits)
+            )
+
+        if isinstance(other, PauliSumOp):
+            return PauliSumOp(
+                self.coeff * self.primitive + other.coeff * other.primitive, coeff=1  # type: ignore
+            )
+
+        from .pauli_op import PauliOp
+
+        if isinstance(other, PauliOp):
+            return PauliSumOp(
+                self.coeff * self.primitive  # type: ignore
+                + other.coeff * SparsePauliOp(other.primitive)
+            )
+
+        return SummedOp([self, other])
+
+    def mul(self, scalar: Union[int, float, complex, ParameterExpression]) -> OperatorBase:
+        if isinstance(scalar, Number) and scalar != 0:
+            return PauliSumOp(scalar * self.primitive, coeff=self.coeff)  # type: ignore
+
+        return super().mul(scalar)
+
+    def adjoint(self) -> OperatorBase:
+        return PauliSumOp(
+            self.primitive.conjugate(), coeff=self.coeff.conjugate()  # type:ignore
+        )
+
+    def equals(self, other: OperatorBase) -> bool:
+        if not isinstance(other, PauliSumOp):
+            return False
+
+        if isinstance(self.coeff, ParameterExpression) or isinstance(
+                other.coeff, ParameterExpression
+        ):
+            return (
+                self.coeff == other.coeff
+                and self.primitive.simplify()  # type:ignore
+                == other.primitive.simplify()  # type:ignore
+            )
+        simple_self = (self.coeff * self.primitive).simplify()  # type:ignore
+        simple_other = (other.coeff * other.primitive).simplify()  # type:ignore
+        return len(simple_self) == len(simple_other) and simple_self == simple_other
```

Comment:
Could it be more efficient to absorb the coefficients into the primitive and use the equality check of the `SparsePauliOp`? Or does that not work because it doesn't simplify?

Reply:
len chack is unnecessary here.
```suggestion
        simple_self = (self.coeff * self.primitive).simplify()  # type:ignore
        simple_other = (other.coeff * other.primitive).simplify()  # type:ignore
        return simple_self == simple_other
```

Reply:
Sorry, I forgot. This is necessary because of broadcasting of Numpy.